spec/jekyll_redirect_from/redirect_page_spec.rb

```diff
@@ -1,148 +1,142 @@
 # frozen_string_literal: true
 
 describe JekyllRedirectFrom::RedirectPage do
-  let(:from) { "/foo" }
-  let(:to) { "/bar" }
   let(:site_url) { site.config["url"] }
-  subject { described_class.from_paths(site, from, to) }
   before { site.read }
 
-  context "being a page" do
-    before { subject.read_yaml(nil, nil, nil) }
+  shared_examples "a redirect page" do
+    context "being a page" do
+      before { subject.read_yaml(nil, nil, nil) }
 
-    it "returns no content" do
-      expect(subject.content).to eql("")
-    end
+      it "returns no content" do
+        expect(subject.content).to eql("")
+      end
 
-    it "returns no output" do
-      expect(subject.output).to eql("")
-    end
+      it "returns no output" do
+        expect(subject.output).to eql("")
+      end
 
-    it "sets default data" do
-      expect(subject.to_liquid["layout"]).to eql("redirect")
-      expect(subject.to_liquid["sitemap"]).to be_falsey
-    end
-  end
+      it "sets default data" do
+        expect(subject.to_liquid["layout"]).to eql("redirect")
+        expect(subject.to_liquid["sitemap"]).to be_falsey
+      end
 
-  context "creating a page from paths" do
-    it "sets the permalink" do
-      expect(subject.to_liquid["permalink"]).to eql(from)
-    end
+      it "sets the paths" do
+        expect(subject.to_liquid["permalink"]).to eql(from)
+        expect(subject.to_liquid).to have_key("redirect")
+        expect(subject.to_liquid["redirect"]["from"]).to eql(from)
+        expect(subject.to_liquid["redirect"]["to"]).to eql("#{site_url}#{to}")
+      end
+
+      it "sets the permalink" do
+        expect(subject.to_liquid["permalink"]).to eql(from)
+      end
 
-    it "sets redirect metadata" do
-      expect(subject.to_liquid).to have_key("redirect")
-      expect(subject.to_liquid["redirect"]["from"]).to eql(from)
-      expect(subject.to_liquid["redirect"]["to"]).to eql("#{site_url}#{to}")
+      it "sets redirect metadata" do
+        expect(subject.to_liquid).to have_key("redirect")
+        expect(subject.to_liquid["redirect"]["from"]).to eql(from)
+        expect(subject.to_liquid["redirect"]["to"]).to eql("#{site_url}#{to}")
+      end
     end
 
-    context "with a document" do
-      let(:doc) { site.documents.first }
+    context "generating" do
+      before { site.generate }
+      let(:output) { Jekyll::Renderer.new(site, subject, site.site_payload).run }
```

Comment:
Not a fan of having new `Renderer` instances for every call to `output` of the same `subject`.
Can't the output be stashed in an instance variable and have that be read on consequent access?

Additionally, I can't find the definition for `subject` in the new diff. Earlier, it was explicit.

Reply:
> new Renderer instances for every call to output
Will check this out.

> I can't find the definition for subject in the new diff
We're using a named subject `subject(:page) { ... }`.

Reply:
> We're using a named subject subject(:page) { ... }.

Again, where is that defined prior to this line?

Reply:
From the `rspec` docs:
> Use let to define a memoized helper method. The value will be cached across multiple calls in the same example but not across examples.

> every call to output of the same subject.

So it's not for _every_ call to `output`, but for every example.

>  Can't the output be stashed in an instance variable and have that be read on consequent access?

@ashmaroli This is possible, but usually goes against idiomatic rspec. Unless it's really causing perf issues, I would want to avoid using instance vars because:
* it is effectively sharing state across examples
* _may_ introduce regressions and cause issues with test ordering at a later point in time.
* `subject` is not accessible outside `let` and `it` blocks, so will have to work around that as well.

Also, it always that way in the test before the refactor as well, so this PR does not "add" this.
Let me know if you still want me to change it to use instance vars.

Reply:
> Again, where is that defined prior to this line?

The snippet that uses `subject` is a part of the shared examples. So it doesn't make assumptions about variable names. Setting `subject` to the right value is the responsibility of the "including" context. In this case, it is set in two places, for each of the two high-level contexts we have:

* Line 72: `subject(:page) { described_class.redirect_to(doc, to) }`
* Line 121: `subject(:page) { described_class.redirect_from(doc, from) }`

Reply:
@ashmaroli Let me know if you want me to use `page` explicitly instead of `subject` inside the shared examples as well.

Reply:
> Let me know if you still want me to change it to use instance vars.

No its fine if the renderer needs to be *different* across examples.
There isn't any noticeable perf issue when running this branch locally. So again, it fine.

> use `page` explicitly instead of `subject` inside the shared examples

If that improves the readability / comprehensibility of the test file, sure. Otherwise, its better to stick with idiomatic expressions to avoid unnecessary confusion.

Reply:
@ashmaroli Cool. I've switched to using a named subject everywhere, including the shared examples (ecedb76).
I've also rebased the branch on upstream's `master`, so it should have the latest fix for Jekyll 3.9.